lib/ui/painting.dart

```diff
@@ -2146,21 +2150,119 @@ Future<Codec> instantiateImageCodecFromBuffer(
+    },
+  );
+}
+
+/// Instantiates an image [Codec].
+///
+/// This method is a convenience wrapper around the [ImageDescriptor] API.
+///
+/// The [buffer] parameter is the binary image data (e.g a PNG or GIF binary data).
+/// The data can be for either static or animated images. The following image
+/// formats are supported: {@macro dart.ui.imageFormats}
+///
+/// The [buffer] will be disposed by this method once the codec has been created,
+/// so the caller must relinquish ownership of the [buffer] when they call this
+/// method.
+///
+/// The [getTargetSize] parameter, when specified, will be invoked and passed
+/// the image's [ImageDescriptor] to determine the size to decode the image to.
+/// The width and the height of the size it returns must be positive values
+/// greater than or equal to 1, or null. It is valid to return a [TargetImageSize]
+/// that specifies only one of `width` and `height` with the other remaining
+/// null, in which case the omitted dimension will be scaled to maintain the
+/// aspect ratio of the original dimensions. When both are null or omitted,
+/// the image will be decoded at its native resolution (as will be the case if
+/// the [getTargetSize] parameter is omitted).
+///
+/// Scaling the image to larger than its intrinsic size should usually be
+/// avoided, since it causes the image to use more memory than necessary.
+/// Instead, prefer scaling the [Canvas] transform.
+///
+/// The returned future can complete with an error if the image decoding has
+/// failed.
+Future<Codec> instantiateImageCodecWithSize(
+  ImmutableBuffer buffer, {
+  TargetImageSizeProducer? getTargetSize,
 }) async {
+  getTargetSize ??= (ImageDescriptor descriptor) => const TargetImageSize();
   final ImageDescriptor descriptor = await ImageDescriptor.encoded(buffer);
-  if (!allowUpscaling) {
-    if (targetWidth != null && targetWidth > descriptor.width) {
-      targetWidth = descriptor.width;
-    }
-    if (targetHeight != null && targetHeight > descriptor.height) {
-      targetHeight = descriptor.height;
-    }
+  try {
+    final TargetImageSize targetSize = getTargetSize(descriptor);
+    assert(targetSize.width == null || targetSize.width! > 0);
+    assert(targetSize.height == null || targetSize.height! > 0);
+    return descriptor.instantiateCodec(
+      targetWidth: targetSize.width,
+      targetHeight: targetSize.height,
+    );
+  } finally {
+    buffer.dispose();
   }
-  buffer.dispose();
-  return descriptor.instantiateCodec(
-    targetWidth: targetWidth,
-    targetHeight: targetHeight,
-  );
+}
+
+/// Signature for a callback that determines the size to which an image should
+/// be decoded given its [ImageDescriptor].
+///
+/// See also:
+///
+///  * [instantiateImageCodecWithSize], which used this signature for its
+///    `getTargetSize` argument.
+typedef TargetImageSizeProducer = TargetImageSize Function(ImageDescriptor descriptor);
```

Comment:
I think we generally use "Callback" as the suffix for function types

Reply:
Done